### PR TITLE
Update to .NET 9 GA. Update some test dependencies.

### LINKED ISSUE #2978: [Feature Request] Update to .NET 9 GA version
Update any .NET 9 dependencies to stable version, update .NET 9 version in build pipeline YAML, fix any new analyzer warnings.

### LINKED ISSUE #3001: [CI] Dotnet tools shouldn't use unconstrained version
**Is your feature request related to a problem? Please describe.**

dotnet-reportgenerator-globaltool is installed in CI to generate a report. But it is installed without a constraint on version. This can lead to build instability.

Installing the dotnet tool in ci should either use dotnet-tools.json or specify a version in the dotnet tool install command.

### PR DESCRIPTION
**Pull request by pmaytak**
Closes #2978
Closes #3001

### CODE REVIEW

File: .github/workflows/dotnetcore.yml [hunk 1]
```diff
@@ -18,7 +18,6 @@ on:
       - dev
 
 env:
-  TargetNet9: True
```
Comment (jmprieur) [inline .github/workflows/dotnetcore.yml:21]: I would actually keep, but rename TargetNet10, as in a few weeks/months, we'll need to add TargetNet10 ...

File: .github/workflows/dotnetcore.yml [hunk 2]
```diff
@@ -18,7 +18,6 @@ on:
...
   TargetNet8: True
```
Comment (jmprieur) [inline .github/workflows/dotnetcore.yml]: Do we need this?

File: benchmark/Microsoft.IdentityModel.Benchmarks/Microsoft.IdentityModel.Benchmarks.csproj [hunk 3]
```diff
@@ -4,8 +4,7 @@
     <AssemblyName>Microsoft.IdentityModel.Benchmarks</AssemblyName>
     <PackageId>Microsoft.IdentityModel.Benchmarks</PackageId>
     <OutputType>Exe</OutputType>
-    <TargetFrameworks>net6.0;net8.0</TargetFrameworks>
```
Comment (jmprieur) [inline benchmark/Microsoft.IdentityModel.Benchmarks/Microsoft.IdentityModel.Benchmarks.csproj:7]: I would keep the whole block, but have for now:

```xml
<TargetFrameworks>net6.0;net8.0;net9.0</TargetFrameworks>
    <TargetFrameworks Condition="'$(TargetNet10)'== 'True'">$(TargetFrameworks);</TargetFrameworks>
```

File: build/targets.props [hunk 4]
```diff
@@ -1,7 +1,6 @@
 <Project>
   <PropertyGroup>
-    <SrcTargets Condition="'$(LocalBuild)' != 'True'">net462;net472;netstandard2.0;net6.0;net8.0</SrcTargets>
-    <SrcTargets Condition="'$(LocalBuild)' == 'True'">netstandard2.0;net8.0</SrcTargets>
-    <SrcTargets Condition="'$(TargetNet9)' == 'True'">$(SrcTargets);net9.0</SrcTargets>
```
Comment (jmprieur) [inline build/targets.props:5]: Same feedback as above. Preserve for TargetNet10.

**Review by jmprieur — APPROVED** [review]
LGTM
I provided some suggestions to save us time in a few weeks/months.

BTW instead of TargetNet10, this could be TargetNetvNext

File: .github/workflows/dotnetcore.yml [hunk 5]
```diff
@@ -30,18 +29,20 @@ jobs:
     - name: Checkout repository
       uses: actions/checkout@v4.1.1
 
-    - name: Setup .NET 6.0.x
+    - name: Setup .NET 6.x
```
Comment (keegan-caruso) [inline .github/workflows/dotnetcore.yml]: 6 and 8 shouldn't be needed I don't think? The 9 sdk can built for 6 and 8.

Comment (pmaytak) [reply .github/workflows/dotnetcore.yml]: Removed.

File: .github/workflows/dotnetcore.yml (hunk 1 above)
Comment (pmaytak) [reply .github/workflows/dotnetcore.yml:21]: Added `TargetNetNext` flag.

File: .github/workflows/dotnetcore.yml (hunk 2 above)
Comment (pmaytak) [reply .github/workflows/dotnetcore.yml]: Not used in this YAML.

File: .github/workflows/dotnetcore.yml [hunk 6]
```diff
@@ -29,19 +28,11 @@ jobs:
     steps:
     - name: Checkout repository
       uses: actions/checkout@v4.1.1
-
-    - name: Setup .NET 6.0.x
```
Comment (msbw2) [inline .github/workflows/dotnetcore.yml:33]: You don't need these anymore?

Comment (msbw2) [reply .github/workflows/dotnetcore.yml:33]: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2990#discussion_r1837050837

**Review by msbw2 — APPROVED** [review]

**Review by westin-m — APPROVED** [review]

File: .github/workflows/aot-check.yml [hunk 7]
```diff
@@ -22,6 +22,11 @@ jobs:
...
         fetch-depth: 1
 
+    - name: Setup .NET 9.0.x
+      uses: actions/setup-dotnet@v4.0.1
+      with:
+        dotnet-version: 9.0.x
+
     - name: Runs powershell script
       id: aot-powershell
       run: build\test-aot.ps1 'net8.0'
```
Comment (msbw2) [inline .github/workflows/aot-check.yml:32]: Target `net9.0`?
```suggestion
      run: build\test-aot.ps1 'net9.0'
```

File: test/Microsoft.IdentityModel.AotCompatibility.TestApp/Microsoft.IdentityModel.AotCompatibility.TestApp.csproj [hunk 8]
```diff
@@ -1,8 +1,8 @@
 <Project Sdk="Microsoft.NET.Sdk">
   
   <PropertyGroup>
-    <TargetFramework>net8.0</TargetFramework>
-    <TargetFrameworks Condition="'$(TargetNet9)' == 'True'">$(TargetFramework); net9.0</TargetFrameworks>
+    <TargetFrameworks>net8.0;net9.0</TargetFrameworks>
```
Comment (msbw2) [inline test/Microsoft.IdentityModel.AotCompatibility.TestApp/Microsoft.IdentityModel.AotCompatibility.TestApp.csproj:4]: There's an error indicating you can only target one of these.
```
C:\Program Files\dotnet\sdk\9.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.CrossTargeting.targets(31,5): error NETSDK1129: The 'Publish' target is not supported without specifying a target framework. The current project targets multiple frameworks, you must specify one of the following frameworks in order to publish: net8.0, net9.0 [D:\a\azure-activedirectory-identitymodel-extensions-for-dotnet\azure-activedirectory-identitymodel-extensions-for-dotnet\test\Microsoft.IdentityModel.AotCompatibility.TestApp\Microsoft.IdentityModel.AotCompatibility.TestApp.csproj]
Push-Location: D:\a\azure-activedirectory-identitymodel-extensions-for-dotnet\azure-activedirectory-identitymodel-extensions-for-dotnet\build\test-aot.ps1:30
```

Comment (msbw2) [reply test/Microsoft.IdentityModel.AotCompatibility.TestApp/Microsoft.IdentityModel.AotCompatibility.TestApp.csproj:4]: Specifically this publish command fails:
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/04d0d137f0b2099ee4aa7fdee8881ce6506f92e0/build/test-aot.ps1#L5

File: .github/workflows/dotnetcore.yml [hunk 9]
```diff
@@ -29,19 +28,11 @@ jobs:
...
-      uses: actions/setup-dotnet@v4.0.0
-      with:
-        dotnet-version: 8.0.x
-    - name: Setup .NET 9.0.x
+        
+    - name: Setup .NET 9.x
       uses: actions/setup-dotnet@v4.0.0
       with:
-        dotnet-version: 9.0.100-rc.2.24474.11
+        dotnet-version: 9.x
```
Comment (msbw2) [inline .github/workflows/dotnetcore.yml:35]: I think this needs to be `9.0.x`:
```suggestion
        dotnet-version: 9.0.x
```

File: .github/workflows/dotnetcore.yml [hunk 10]
```diff
@@ -29,19 +28,11 @@ jobs:
...
 
     - name: Run the tests
       run: dotnet test Wilson.sln --collect:"XPlat Code Coverage" --settings:./build/CodeCoverage.runsettings
```
Comment (msbw2) [inline .github/workflows/dotnetcore.yml:42]: To fix #3001 update the line a few lines below:
`dotnet tool install -g dotnet-reportgenerator-globaltool --version 5.4.0`

https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/1a9f885dd8fd7bdcfc9cd27064884e5254abb81b/.github/workflows/dotnetcore.yml#L42

File: test/Microsoft.IdentityModel.AotCompatibility.TestApp/Microsoft.IdentityModel.AotCompatibility.TestApp.csproj (hunk 8 above)
Comment (jennyf19) [reply test/Microsoft.IdentityModel.AotCompatibility.TestApp/Microsoft.IdentityModel.AotCompatibility.TestApp.csproj:4]: It needs the target framework at the end `-f net8.0`

File: .github/workflows/dotnetcore.yml (hunk 9 above)
Comment (pmaytak) [reply .github/workflows/dotnetcore.yml:35]: Both work.

File: .github/workflows/aot-check.yml (hunk 7 above)
Comment (pmaytak) [reply .github/workflows/aot-check.yml:32]: I was trying to keep this the same and still target net8 here.

**Pull request merged**